src/receive_buffer.cpp

```diff
@@ -41,44 +41,66 @@ namespace libtorrent {
 		}
 	}
 
-int receive_buffer::max_receive()
+int receive_buffer::max_receive() const
 {
-	int max = packet_bytes_remaining();
-	if (m_recv_pos >= m_soft_packet_size) m_soft_packet_size = 0;
-	if (m_soft_packet_size && max > m_soft_packet_size - m_recv_pos)
-		max = m_soft_packet_size - m_recv_pos;
-	return max;
+	return int(m_recv_buffer.size() - m_recv_end);
 }
 
 boost::asio::mutable_buffer receive_buffer::reserve(int size)
 {
 	TORRENT_ASSERT(size > 0);
 	TORRENT_ASSERT(m_recv_pos >= 0);
-	// this is unintuitive, but we used to use m_recv_pos in this function when
-	// we should have used m_recv_end. perhaps they always happen to be equal
-	TORRENT_ASSERT(m_recv_pos == m_recv_end);
 
 	// normalize() must be called before receiving more data
 	TORRENT_ASSERT(m_recv_start == 0);
 
-	m_recv_buffer.resize(m_recv_end + size);
+	if (m_recv_buffer.size() < m_recv_end + size)
+	{
+		// TODO: 4 if the message size is larger still, allocate enough space to
+		// fit the whole message
+		buffer new_buffer(m_recv_end + size
+			, aux::array_view<char const>(m_recv_buffer.ptr(), m_recv_end));
+		m_recv_buffer = std::move(new_buffer);
+
+		// since we just increased the size of the buffer, reset the watermark to
+		// start at our new size (avoid flapping the buffer size)
+		m_watermark = sliding_average<20>();
+	}
+
 	return boost::asio::buffer(&m_recv_buffer[0] + m_recv_end, size);
 }
 
+void receive_buffer::grow(int const limit)
+{
+	int const current_size = int(m_recv_buffer.size());
+	TORRENT_ASSERT(current_size < std::numeric_limits<int>::max() / 3);
+
+	// first grow to one piece message, then grow by 50% each time
+	int const new_size = std::min((current_size < m_packet_size)
+		? m_packet_size : current_size * 3 / 2, limit);
```

Comment:
yeah, I was going to restrict what limit can be set to. that was my original plan at least. the thing is that m_packet_size is read from the stream, so it kind of makes sense to still apply the limit to it I think